### ISSUE TITLE
Wrong type hit when transformer is provided

### ISSUE THREAD

**Issue by majkrzak** [lark-parser/lark#1186]
When `transformer` is provided to the `Parser` constructor, the `parse` method may no longer return the parse `'ParseTree'`

**Comment by erezsh** [comment]
Yes, that's true. Ideally `parse` should take a generic based on the transformer parameter (if given), and return its return value.

**Comment by yeungchie** [comment]
I'm running into the same issue. Is there currently a recommended approach? 
I'm handling it in one of these ways, but maybe there's something better:

1. Runtime assertion:

```python
def parse_my_object(text: str) -> MyObject:
    ...
    a = Lark(...).parse(text)
    assert isinstance(a, MyObject)
    return a
```

2. Using `typing.cast`:

```python
a = cast(MyObject, Lark(...).parse(text))
```

**Comment by erezsh** [comment]
Linked PR should fix it
